### PR TITLE
Restore CacheFileRegion refcounting for writes

### PR DESCRIPTION
**Pull request by tlrx**
In #98241 we removed the refcounting around read and write handler in SharedBytes.IO. But recently we saw wrong bytes being read from the snapshot file under heavy evictions and investigation shows that the bytes belonged to another cached file.

Low level logging (hard to reproduce) show that writes and reads using the same `SharedBytes.IO` instance but for different cached files could be interleaved, so that bytes in shared cache could be overwritten and the last read would read (and store in internal index input buffers) bytes from a different file:

```
Thread[elasticsearch[node_t0][stateless_shard][T#4],5,TGRP-IndexCorruptionIT]: 10485760 bytes written using SharedBytes$IO@dc07632 (230716978) for FileCacheKey[shardId=[index-0][0], primaryTerm=1, fileName=stateless_commit_26]
Thread[elasticsearch[node_t0][stateless_shard][T#3],5,TGRP-IndexCorruptionIT]: 10485760 bytes written using  SharedBytes$IO@dc07632 (230716978) for FileCacheKey[shardId=[index-0][0], primaryTerm=1, fileName=stateless_commit_16]
Thread[elasticsearch[node_t0][stateless_shard][T#4],5,TGRP-IndexCorruptionIT]: 375 bytes read using SharedBytes$IO@dc07632 (230716978) for key FileCacheKey[shardId=[index-0][0], primaryTerm=1, fileName=stateless_commit_26]
```

This pull request changes the read handler so that the IO instance is decref after bytes are read so that the IO instance does not return to the free regions pool too early. It also hold a ref until the writes are done for safety.

Relates #98241

### CODE REVIEW

**Review by original-brownbear — COMMENTED** [review]
As just explained on Slack, not sure this will help.
Quoting myself:

The logic behind try-read uses optimistic concurrency and checks that a page was still owned after a read, so it could technically (and likely will) repeatedly read the wrong bytes but that shouldn't cause trouble (unless there's a bug in the reasoning about the order of things that is).

**Review by kingherc — COMMENTED** [review]
Not an expert in this code to say definitely what's wrong, but just a couple of comments from taking a quick look

**Comment by tlrx** [comment]
Armin an I discussed this change via another channel and we both suspect that the important bit is adressed in https://github.com/elastic/elasticsearch/pull/102848

I'm closing this until we evaluate the results of the other change and will reopen if needed.

**Comment by tlrx** [comment]
Reopening as we think it is needed.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -594,6 +594,11 @@ public final boolean isEvicted() {
         }
     }
 
+    private static final AlreadyClosedException ALREADY_CLOSED_EXCEPTION = new AlreadyClosedException(
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:597]: Any reason we need to be this tricky here? I get that throwing this thing has some cost to it, but for debug purposes it's actually kinda nice to have it and it should be very rare to run into this in production so the performance of this thing shouldn't matter at all?

[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:597]: Note that I'm OK to revert if you think having the stacktrace can help, but I must admit I never found it useful here 😅

**Review by original-brownbear — APPROVED** [review]
LGTM though I'd revert the exception magic :) to keep things less magical. Thanks for finding this Tanguy!

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java]: Sure, I'm fine with reverting.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 2]
```diff
@@ -709,24 +714,33 @@ void populateAndRead(
...
+                            // we rethrow the same instance of AlreadyClosedException
+                            throw ALREADY_CLOSED_EXCEPTION;
+                        }
+                        try {
+                            final int start = Math.toIntExact(gap.start());
+                            var ioRef = io;
+                            assert regionOwners.get(ioRef) == cacheFileRegion;
+                            writer.fillCacheRange(
+                                ioRef,
+                                start,
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:727]: I think this part changed and offsets are now adjusted at a lower level. Anyway, this is not what this change tries to fix so I'll let this unchanged.

**Comment by tlrx** [comment]
Thanks a lot Armin & Iraklis! Just in time for 8.12.0.

**Pull request merged**